src/main/java/org/cbioportal/cdd/repository/topbraid/TopBraidRepository.java

```diff
@@ -0,0 +1,108 @@
+ * be liable to any party for direct, indirect, special,
+ * incidental or consequential damages, including lost profits, arising
+ * out of the use of this software and its documentation, even if
+ * Memorial Sloan-Kettering Cancer Center
+ * has been advised of the possibility of such damage.
+*/
+
+package org.cbioportal.cdd.repository.topbraid;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+
+import org.apache.log4j.Logger;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
+import org.springframework.stereotype.Repository;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+/**
+ *
+ * @author Manda Wilson
+ **/
+@Repository
+public abstract class TopBraidRepository<T> {
+
+    private final static Logger logger = Logger.getLogger(TopBraidRepository.class);
+
+    @Value("${topbraid.url}")
+    private String topBraidURL;
+
+    @Autowired
+    private TopBraidSessionConfiguration topBraidSessionConfiguration;
+
+    protected List<T> query(String query, ParameterizedTypeReference<List<T>> parameterizedType)
+            throws TopBraidException {
+        return query(query, parameterizedType, true);
+    }
+
+    private List<T> query(String query, ParameterizedTypeReference<List<T>> parameterizedType, boolean refreshSessionOnFailure)
+            throws TopBraidException {
+        logger.debug("query() -- query: '" + query + "'");
+        String sessionId = topBraidSessionConfiguration.getSessionId();
+        logger.debug("query() -- sessionId: " + sessionId);
+        RestTemplate restTemplate = new RestTemplate();
+
+        // the default supported types for MappingJackson2HttpMessageConverter are:
+        //   application/json and application/*+json
+        // our response content type is application/sparql-results+json-simple
+        // NOTE: if the response content type was one of the default types we 
+        //   would not have to add the message converter to the rest template
+        MappingJackson2HttpMessageConverter messageConverter = new MappingJackson2HttpMessageConverter();
+        messageConverter.setSupportedMediaTypes(Collections.singletonList(
+            new MediaType("application","sparql-results+json-simple")));
+        restTemplate.getMessageConverters().add(messageConverter);
```

Comment:
Isn't this adding a messageConverter every time we call in with a query?  Should we be checking if the converter is already available before adding it? Or maybe adding it earlier during system startup (SpringBootServletInitializer?)

Reply:
@sheridancbio Spring doesn't support sparql by default - [here are the available types](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/converter/HttpMessageConverter.html)

- AbstractGenericHttpMessageConverter
-  AbstractHttpMessageConverter
-  AbstractJackson2HttpMessageConverter
-  AbstractJaxb2HttpMessageConverter
-  AbstractJsonHttpMessageConverter
-  AbstractWireFeedHttpMessageConverter
-  AbstractXmlHttpMessageConverter
-  AllEncompassingFormHttpMessageConverter
-  AtomFeedHttpMessageConverter
-  BufferedImageHttpMessageConverter
-  ByteArrayHttpMessageConverter
-  FormHttpMessageConverter
-  GsonHttpMessageConverter
-  Jaxb2CollectionHttpMessageConverter
-  Jaxb2RootElementHttpMessageConverter
-  JsonbHttpMessageConverter
-  MappingJackson2CborHttpMessageConverter
-  MappingJackson2HttpMessageConverter
-  MappingJackson2SmileHttpMessageConverter
-  MappingJackson2XmlHttpMessageConverter
-  MarshallingHttpMessageConverter
-  ObjectToStringHttpMessageConverter
-  ProtobufHttpMessageConverter
-  ProtobufJsonFormatHttpMessageConverter
-  ResourceHttpMessageConverter
-  ResourceRegionHttpMessageConverter
-  RssChannelHttpMessageConverter
-  SourceHttpMessageConverter
-  StringHttpMessageConverter

Reply:
Right .. we have to add it .. I'm just thinking that by having the call inside of the function query() we may be adding additional redundant converters to restTemplate during every future call to this function.

Reply:
@sheridancbio I tested moving this out and it doesn't work -- restTemplate.getMessageConverters().add(messageConverter); adds the new message converter to this instance of 'restTemplate' only, not to a more global list of message converters.